### PR TITLE
Notebook2 login

### PR DESCRIPTION
**Pull request by akotlar**
Stacks on #5430. Once #5430 is in, the changes here will be limited to: 1) notebook.py: login/logout routes, the provision of authorized users, auth0 lib, 2) index.html 3) header.html: update lines 12 and 13 to read user from session.

Provides basic login page. Below are a few images of it in action. Looks like app.hail.is. Handles authorized and workshop-only login. Handles login only; future PR will extend to checking, refreshing the session.

cc @cseed 

screenshots (notebook create button not yet PR'd , auth0 page not yet styled)

screen shot 2019-02-25 at 11 17 37 pm
screen shot 2019-02-25 at 11 18 00 pm
screen shot 2019-02-25 at 11 18 18 pm
screen shot 2019-02-25 at 11 18 33 pm

### CODE REVIEW

File: notebook2/notebook/notebook.py [hunk 1]
```diff
@@ -315,6 +347,51 @@ def wait_websocket(ws):
     log.info(f'notification sent to user for {svc_name} {pod_name}')
 
 
+@app.route('/auth0-callback')
+def auth0_callback():
+    # https://github.com/auth0-samples/auth0-python-web-app/commit/d048d6497caa714c52e8411a5f37500787e37305
+    auth0.authorize_access_token()
```
Comment (danking) [inline notebook2/notebook/notebook.py]: It seems like the appropriate link would be to the auth0 docs: https://auth0.com/docs/quickstart/webapp/python, but, IMO, we can expect readers of this code to familiarize themselves with the basic docs of the libraries we use.

File: notebook2/notebook/notebook.py [hunk 2]
```diff
@@ -315,6 +347,51 @@ def wait_websocket(ws):
...
+
+
+@app.route('/login', methods=['GET'])
+def login_page():
+    return render_template('login.html', unauthorized = request.args.get('unauthorized'))
+
+
+@app.route('/login', methods=['POST'])
+def login_auth0():
+    # FIXME ?: Could be placed outside route
```
Comment (danking) [inline notebook2/notebook/notebook.py]: doesn't much matter to me. I tend to avoid optimizing my code until necessary.

Comment (danking) [reply notebook2/notebook/notebook.py]: Either way, we should make a decision and remove FIXME.

File: notebook2/notebook/notebook.py [hunk 3]
```diff
@@ -315,6 +347,51 @@ def wait_websocket(ws):
...
+    external_url = flask.url_for('auth0_callback', _external = True)
+    session['workshop_password'] = request.form.get('workshop-password')
+
+    return auth0.authorize_redirect(redirect_uri = external_url, audience = f'{AUTH0_BASE_URL}/userinfo', prompt = 'login')
+
+
+@app.route('/logout')
```
Comment (danking) [inline notebook2/notebook/notebook.py]: logout feels like a `POST`, no? It's not idempotent.

File: notebook2/notebook/notebook.py [hunk 4]
```diff
@@ -315,6 +347,51 @@ def wait_websocket(ws):
...
+    auth0.authorize_access_token()
+
+    userinfo = auth0.get('userinfo').json()
+
+    email = userinfo['email']
+    workshop_password = session['workshop_password']
+    del session['workshop_password']
+
+    if AUTHORIZED_USERS.get(email) is None and workshop_password != PASSWORD:
+        return redirect(flask.url_for('login_page', unauthorized = True))
```
Comment (danking) [inline notebook2/notebook/notebook.py]: It seems a bit odd to pun the login page as both a login page and an unauthorized page. I realize there's some HTML duplication, but, code-structure-wise, it seems much clearer to just have a `/unauthorized` page. What do you think?

File: notebook2/notebook/notebook.py [hunk 5]
```diff
@@ -58,11 +62,36 @@ def read_string(f):
         return f.read().strip()
 
 KUBERNETES_TIMEOUT_IN_SECONDS = float(os.environ.get('KUBERNETES_TIMEOUT_IN_SECONDS', 5.0))
-app.secret_key = read_string('/notebook-secrets/secret-key')
+
+AUTHORIZED_USERS = read_string('/notebook-secrets/authorized-users').split(',')
```
Comment (danking) [inline notebook2/notebook/notebook.py]: This does not currently exist in the `notebook-secrets` secret. We need to update that secret or create a new secret and update the deployment.yaml.in appropriately.

File: notebook2/Dockerfile [hunk 6]
```diff
@@ -15,7 +15,9 @@ RUN apk add \
   flask \
   Flask_Sockets \
   kubernetes \
-  'urllib3<1.24'
+  'urllib3<1.24'\
+  libsass \
```
Comment (danking) [inline notebook2/Dockerfile]: ah yeah, so currently we don't test notebook/notebook2 automatically.

I usually test my PRs by creating a temporary k8s namespace, and modifying the yaml to deploy into said namespace, that way I can ensure it all starts up correctly. This also, of course, requires putting the necessary secrets and serviceaccounts into said namespace. The configuration for the service accounts are stored in `vdc/k8s-config.yaml`. For the secret, I manually regenerate it using `kubectl create secret generic`

File: notebook2/notebook/notebook.py (hunk 1 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: This was mostly a comment to show users how to extract the access token, which is effectively undocumented.I found it not at all clear, from either auth lib or auth0. In general, I’ve found the reliance on a 3rd party package (relative to auth0) unpleasant (for instance, try changing the request_type to something other than ‘code’).

File: notebook2/notebook/notebook.py (hunk 3 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: I think it is idempotent? If you log out twice you clear the session twice, and issue two logout calls to auth0; the 2nd is ignored.

File: notebook2/notebook/notebook.py (hunk 4 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: I saw it as an action that originated on the login page, and updates that page’s state after completion (so login page manages both submission and handles the response).

I can see your point, happy to create an unauthorized page.

File: notebook2/Dockerfile (hunk 6 above)
Comment (akotlar) [reply notebook2/Dockerfile]: Got it, thanks!

File: notebook2/notebook/notebook.py [hunk 7]
```diff
@@ -58,11 +62,36 @@ def read_string(f):
...
 
+app.config.update(
+    SECRET_KEY = read_string('/notebook-secrets/secret-key'),
+    SESSION_COOKIE_SAMESITE = 'lax',
+    SESSION_COOKIE_HTTPONLY = True,
+    SESSION_COOKIE_SECURE = os.environ.get("NOTEBOOK_DEBUG") != "1"
+)
+
+AUTH0_CLIENT_ID = 'Ck5wxfo1BfBTVbusBeeBOXHp3a7Z6fvZ'
+AUTH0_BASE_URL = 'https://hail.auth0.com'
```
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Sure I’ll look into that.

File: notebook2/notebook/notebook.py (hunk 2 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Yeah I really don’t like unnecessary assignments/regeneration of a constant right value, but I wanted your opinion.

I’ll remove

File: notebook2/notebook/notebook.py (hunk 5 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Got it

File: notebook2/notebook/notebook.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Could you help me understand what this file is used for, and on which server it's stored?

File: notebook2/notebook/notebook.py (hunk 4 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Done, /error endpoint added

File: notebook2/notebook/notebook.py (hunk 3 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Moved to POST, does feel more like a POST (well, DELETE, but that can't be submitted by an HTML form)

File: notebook2/notebook/notebook.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/notebook.py]: As discussed in person, cotton has admin privileges to the auth0 account, and we can add any other team members. If we also want to store usernames/passwords to auth0 in credentials.txt we can do that, at some small security cost (rotating passwords will be slightly more involved, slightly larger chance pass leaks)

File: notebook2/notebook/notebook.py (hunk 1 above)
Comment (danking) [reply notebook2/notebook/notebook.py]: Maybe we're talking about different things? The auth0 quickstart clearly documents this pattern. Which method has a `request_type` parameter? I don't see that in the authlib docs.

File: notebook2/notebook/notebook.py (hunk 3 above)
Comment (danking) [reply notebook2/notebook/notebook.py]: Yeah, from a very practical point of view, we don't want the browser pre-fetching `/logout` ;).

File: notebook2/notebook/notebook.py [hunk 8]
```diff
@@ -315,6 +347,55 @@ def wait_websocket(ws):
...
+        'email': email,
+        'picture': userinfo['picture'],
+    }
+
+    return redirect('/')
+
+
+@app.route('/error', methods=['GET'])
+def error_page():
+    return render_template('error.html', error = request.args.get('err'))
```
Comment (danking) [inline notebook2/notebook/notebook.py]: this template isn't checked in.

File: notebook2/notebook/notebook.py (hunk 1 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: I had wanted to document how to extract out the access token from the response, which I don't think is documented. For the grant that we're using it isn't really necessary, so I'll drop the comment.

File: notebook2/notebook/notebook.py (hunk 3 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Yeah, that would be annoying. I need to switch thinking about html elements as convenient elements to bind event handlers.

File: notebook2/notebook/notebook.py (hunk 8 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Oops!

**Pull request merged**